File: 250723/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const grid = input.slice(1, n + 1).map(line => line.split(' ').map(Number));
const visited = Array.from({length: n}, () => Array(m).fill(false));
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];
function bfs(x, y){
    const queue = [[x, y]];
    visited[y][x] = true;

    while(queue.length > 0){
        const [cx, cy] = queue.shift();

        for(let i = 0; i <4; i++){
            const nx = cx + dx[i];
            const ny = cy + dy[i];

            if(0<= nx && nx < m && 0 <= ny && ny < n && !visited[ny][nx] && grid[ny][nx] === 1){
                visited[ny][nx] = true;
                queue.push([nx, ny]);
                if(nx === m-1 && ny === n-1){
                    return true;
                }
            }
        }
    } 
    return false
}
const flag = bfs(0, 0);

console.log(flag ? 1 : 0)

File: 250821/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.py
from collections import deque

n, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]
visited = [[False for _ in range(m)] for _ in range(n)]


dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

q = deque()

def bfs():
    while q:
        x, y = q.popleft()

        for i in range(4):
            curX = dx[i] + x
            curY = dy[i] + y
            if 0 <= curX and curX < m and 0<= curY \
            and curY < n and grid[curY][curX] == 1 \
            and not visited[curY][curX]:
                visited[curY][curX] = True
                q.append((curX, curY))


q.append((0, 0))
visited[0][0] = True
bfs()

print(1 if visited[n-1][m-1] else 0)
